pkg/plugins/manager_test.go

```diff
@@ -0,0 +1,145 @@
+package plugins
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/go-plugin"
+)
+
+const registeredPluginName = "plugin"
+
+var basicHS = PluginHandshake{
+	Key:   "key",
+	Value: "value",
+}
+
+type testCase struct {
+	m           *Manager[PingPongStub]
+	name        string
+	p           plugin.Plugin
+	expectedErr error
+	description string
+}
+
+type testCases []testCase
+
+func TestManager_LoadPluginClient(t *testing.T) {
+	id := PluginIdentity{
+		ProtocolVersion: 1,
+	}
+	cases := testCases{
+		{
+			NewManager[PingPongStub](),
+			registeredPluginName,
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			nil,
+			"successful registration",
+		},
+		{
+			&Manager[PingPongStub]{},
+			registeredPluginName,
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			ErrUninitializedManager,
+			"uninitialized manager",
+		},
+		{
+			NewManager[PingPongStub](),
+			"someNonExistingPlugin",
+			GRPCPlugin{Impl: &PingPongPlayer{}},
+			ErrPluginNameNotFound,
+			"unregistered service",
+		},
+		{
+			NewManager[PingPongStub](),
+			registeredPluginName,
+			NopGRPCPlugin{},
+			ErrPluginOfWrongType,
+			"wrong type plugin",
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.description, func(t *testing.T) {
+			// Necessary to redefine in every test due to future setting of stdout by the go-plugin package
+			id.Cmd = *pluginServerCmd(id.ProtocolVersion, basicHS)
+			if tc.p != nil {
+				tc.m.RegisterPlugin(registeredPluginName, id, basicHS, tc.p)
+			}
+			_, err := tc.m.LoadPluginClient(tc.name)
+			assertErr(t, err, tc.expectedErr, tc.description)
+		})
+	}
+}
+
+func assertErr(t *testing.T, err, cmpErr error, desc string) {
+	if err != nil && !errors.Is(err, cmpErr) {
+		t.Errorf("'%s' failed: %v", desc, err)
+	}
+}
+
+// Used to run the plugin server
+func pluginServerCmd(version uint, auth PluginHandshake) *exec.Cmd {
+	cs := []string{"-test.run=TestPluginServer", "--"}
+	cs = append(cs, auth.Key, auth.Value, strconv.Itoa(int(version)))
+	cmd := exec.Command(os.Args[0], cs...)
+
+	env := []string{
+		"RUN_PLUGIN_SERVER=1",
+	}
+	cmd.Env = append(env, os.Environ()...)
+	return cmd
+}
+
+// This is not a real test. This is the plugin server that will be triggered by the tests
+func TestPluginServer(*testing.T) {
```

Comment:
If this is not a test, it should start with something other than `Test`.  Also not sure why it takes `*testing.T` parameter.

Reply:
It's not a test, yet the command that runs it is `go test` which accepts the `-test.run` flag that specifies the tests it should run. Unfortunately, golang don't allow tests to be named other than `Test...` and must accept `*testing.T` as an argument:

> You write a test by creating a file with a name ending in `_test.go` that contains functions named `TestXXX` with signature `func (t *testing.T)`.

Reply:
I like the ping-pong protobuf and additional classes, since it explains very well how plugins and grcp work. However, I'm wondering about this methodology, because this test doesn't (and shouldn't) test the logic of the plugins. I would expect this test to mock both `plugin.Plugin` and `plugin.Client`, so it can become a strict unit test.

One problem is the `plugin.Client` is a struct and not an interface, but I suggest creating an interface on our side in order to allow mocking it as well:

```golang
type Client interface {
    Client() (ClientProtocol, error)
}
```